File: binarytree/binarytree.ts
class Leaf {
  public key: number;
  public name: string;

  public leftChild: Leaf;
  public rightChild: Leaf;

  constructor(key: number, name: string) {
    this.key = key;
    this.name = name;

    this.leftChild = null;
    this.rightChild = null;
  };

  public toString(): string {
    return this.name + " has a key of " + this.key;
  };
};

class BinaryTree {
  public root: Leaf;

  constructor() {
    this.root = null;
  };

  public addLeaf(key: number, name: string) {
    const newLeaf = new Leaf(key, name);

    if (this.root === null) {
      this.root = newLeaf;
    } else {
      let parentLeaf: Leaf;
      let focusLeaf: Leaf = this.root;
      while (true) {
        parentLeaf = focusLeaf;

        if (key < focusLeaf.key) {
          focusLeaf = focusLeaf.leftChild;

          if (focusLeaf === null) {
            parentLeaf.leftChild = newLeaf;
            return;
          }
        } else {
          focusLeaf = focusLeaf.rightChild;

          if (focusLeaf === null) {
            parentLeaf.rightChild = newLeaf;
            return;
          }
        }
      }
    }
  };

  public findLeaf(key: number): Leaf {
    let focusLeaf: Leaf = this.root;

    while (focusLeaf.key != key) {
      if (key < focusLeaf.key) {
        focusLeaf = focusLeaf.leftChild;
      } else {
        focusLeaf = focusLeaf.rightChild;
      }

      if (focusLeaf == null) {
        return null;
      }
    }

    return focusLeaf;
  };

  public getReplacementLeaf(leaf: Leaf): Leaf {
    let replacementParent = leaf;
    let replacementLeaf = leaf;

    let focusNode = leaf.rightChild;

    while (focusNode !== null) {
      replacementParent = replacementLeaf;
      replacementLeaf = focusNode;
      focusNode = focusNode.leftChild;
    }

    if (replacementLeaf !== leaf.rightChild) {
      replacementParent.leftChild = replacementLeaf.rightChild;
      replacementLeaf.rightChild = leaf.rightChild;
    }

    return replacementLeaf;
  };

  public remove(key: number): boolean {
    let parent = this.root;
    let focusLeaf = this.root;

    let isItALeftChild: boolean = true;
    while(focusLeaf.key !== key) {
      parent = focusLeaf;

      if (key < focusLeaf.key) {
        isItALeftChild = true;
        focusLeaf = focusLeaf.leftChild;
      } else {
        isItALeftChild = false;
        focusLeaf = focusLeaf.rightChild;
      }

      if (focusLeaf === null) {
        return false;
      }
    }

    // no children
    if (focusLeaf.leftChild == null && focusLeaf.rightChild == null) {
      if (focusLeaf == this.root) {
        this.root = null;
      } else if (isItALeftChild) {
        parent.leftChild = null;
      } else {
        parent.rightChild = null;
      }
    }

    // no right child
    else if(focusLeaf.rightChild == null) {
      if (focusLeaf == this.root) {
        this.root = focusLeaf.leftChild;
      } else if (isItALeftChild) {
        parent.leftChild = focusLeaf.leftChild;
      } else {
        parent.rightChild = focusLeaf.leftChild;
      }
    }

    // no left child
    else if (focusLeaf.leftChild == null) {
      if (focusLeaf == this.root) {
        this.root = focusLeaf.rightChild;
      } else if (isItALeftChild) {
        parent.leftChild = focusLeaf.rightChild;
      } else {
        parent.rightChild = focusLeaf.rightChild;
      }
    }

    // has two children
    else {
      let replacementLeaf = this.getReplacementLeaf(focusLeaf);

      if (focusLeaf == this.root) {
        this.root = replacementLeaf;
      } else if (isItALeftChild) {
        parent.leftChild = replacementLeaf;
      } else {
        parent.rightChild = replacementLeaf;
      }

      replacementLeaf.leftChild = focusLeaf.leftChild;
    }

    return true;
  };

  public inOrderTraverseTree(focusLeaf: Leaf): void {
    if (focusLeaf !== null) {
      this.inOrderTraverseTree(focusLeaf.leftChild);
      console.log(focusLeaf.toString());
      this.inOrderTraverseTree(focusLeaf.rightChild);
    }
  };

  public preOrderTraverseTree(focusLeaf: Leaf): void {
    if (focusLeaf !== null) {
      console.log(focusLeaf.toString());
      this.preOrderTraverseTree(focusLeaf.leftChild);
      this.preOrderTraverseTree(focusLeaf.rightChild);
    }
  };

  public postOrderTraverseTree(focusLeaf: Leaf): void {
    if (focusLeaf !== null) {
      this.postOrderTraverseTree(focusLeaf.leftChild);
      this.postOrderTraverseTree(focusLeaf.rightChild);
      console.log(focusLeaf.toString());
    }
  };

  public reverse(focusLeaf: Leaf): void {
    if (focusLeaf !== null) {
      const temp = focusLeaf.leftChild;
      focusLeaf.leftChild = focusLeaf.rightChild;
      focusLeaf.rightChild = temp;

      this.reverse(focusLeaf.leftChild);
      this.reverse(focusLeaf.rightChild);
    }
  };
}

// main
(() => {
  const binaryTree: BinaryTree = new BinaryTree();

  binaryTree.addLeaf(50, "Jim");
  binaryTree.addLeaf(25, "James");
  binaryTree.addLeaf(15, "Joe");
  binaryTree.addLeaf(30, "Jeff");
  binaryTree.addLeaf(75, "Jarry");
  binaryTree.addLeaf(85, "Joel");

  binaryTree.inOrderTraverseTree(binaryTree.root);
  binaryTree.reverse(binaryTree.root);
  console.log("***********");
  binaryTree.inOrderTraverseTree(binaryTree.root);

  // binaryTree.inOrderTraverseTree(binaryTree.root);
  // binaryTree.preOrderTraverseTree(binaryTree.root);
  // binaryTree.postOrderTraverseTree(binaryTree.root);

  // console.log(binaryTree.findLeaf(15).toString());

  // binaryTree.preOrderTraverseTree(binaryTree.root);
  // console.log("************");
  // binaryTree.remove(15);
  // binaryTree.preOrderTraverseTree(binaryTree.root);
  // console.log("************");
  // binaryTree.remove(25);
  // binaryTree.preOrderTraverseTree(binaryTree.root);
  // console.log("************");
  // binaryTree.remove(75);
  // binaryTree.preOrderTraverseTree(binaryTree.root);
  // console.log("************");
  // binaryTree.remove(50);
  // binaryTree.preOrderTraverseTree(binaryTree.root);
})();

File: java/src/main/java/binarytree/BinaryTree.java
package binarytree;

class Node {
    public int key;
    public String data;

    public Node leftChild = null;
    public Node rightChild = null;

    public Node(int key, String data) {
        this.key = key;
        this.data = data;
    };
}

public class BinaryTree {
    public Node root;

    public BinaryTree () {
        this.root = null;
    }

    public boolean addNode(int key, String data) {
        Node newNode = new Node(key, data);

        if (this.root == null) {
            this.root = newNode;
            return true;
        }

        Node parentNode;
        Node focusNode = this.root;
        while (true) {
            parentNode = focusNode;

            if (key < focusNode.key) {
                focusNode = focusNode.leftChild;

                if (focusNode == null) {
                    parentNode.leftChild = newNode;
                    return true;
                }
            } else {
                focusNode = focusNode.rightChild;

                if (focusNode == null) {
                    parentNode.rightChild = newNode;
                    return true;
                }
            }
        }
    }

    public void findNode(int key) {
        Node focusNode = this.root;

        while (focusNode.key != key) {
            if (focusNode.key > key) {
                focusNode = focusNode.leftChild;
            } else {
                focusNode = focusNode.rightChild;
            }

            if (focusNode == null) {
                System.out.println("null");
                return;
            }
        }

        System.out.println(focusNode.key + " " + focusNode.data);
    }

    public Node getReplacementNode(Node replacedNode) {
        Node replacementParentNode = replacedNode;
        Node replacementNode = replacedNode;

        Node focusNode = replacedNode.rightChild;
        while (focusNode != null) {
            replacementParentNode = replacementNode;
            replacementNode = focusNode;
            focusNode = focusNode.leftChild;
        }

        if (replacementNode != replacedNode.leftChild) {
            replacementParentNode.leftChild = replacementNode.rightChild;
            replacementNode.rightChild = replacedNode.rightChild;
        }

        return replacementNode;
    }

    // delete node
    public boolean remove(int key) {
        Node parentNode = this.root;
        Node focusNode = this.root;

        boolean isLeftChild = true;
        while (focusNode.key != key) {
            parentNode = focusNode;

            isLeftChild = key < focusNode.key;
            if (isLeftChild) {
                focusNode = focusNode.leftChild;
            } else {
                focusNode = focusNode.rightChild;
            }

            if (focusNode == null) {
                return false;
            }
        }

        // no children
        if (focusNode.leftChild == null && focusNode.rightChild == null) {
            if (focusNode == this.root) {
                this.root = null;
            } else if (isLeftChild) {
                parentNode.leftChild = null;
            } else {
                parentNode.rightChild = null;
            }
        }

        // no right child
        else if (focusNode.rightChild == null) {
            if (focusNode == this.root) {
                this.root = focusNode.leftChild;
            } else if (isLeftChild) {
                parentNode.leftChild = focusNode.leftChild;
            } else {
                parentNode.rightChild = focusNode.leftChild;
            }
        }

        // no left child
        else if (focusNode.leftChild == null) {
            if (focusNode == this.root) {
                this.root = focusNode.rightChild;
            } else if (isLeftChild) {
                parentNode.leftChild = focusNode.rightChild;
            } else {
                parentNode.rightChild = focusNode.rightChild;
            }
        }

        // has children
        else {
            Node replacementNode = this.getReplacementNode(focusNode);

            if (focusNode == this.root) {
                this.root = replacementNode;
            } else if (isLeftChild) {
                parentNode.leftChild = replacementNode;
            } else {
                parentNode.rightChild = replacementNode;
            }

            replacementNode.leftChild = focusNode.leftChild;
        }

        return true;
    }

    public void reverse(Node node) {
        if (node != null) {
            Node tempNode = node.leftChild;
            node.leftChild = node.rightChild;
            node.rightChild = tempNode;

            this.reverse(node.leftChild);
            this.reverse(node.rightChild);
        }
    }

    public Node findNodeWithClosestKey(int key) {
        Node closestNode = this.root;
        Node focusNode = this.root;

        int diff = Math.abs(focusNode.key - key);
        while (focusNode != null) {
            int newDiff = Math.abs(focusNode.key - key);
            if (diff > newDiff) {
                diff = newDiff;
                closestNode = focusNode;
            }

            if (focusNode.key > key) {
                focusNode = focusNode.leftChild;
            } else {
                focusNode = focusNode.rightChild;
            }
        }

        return closestNode;
    }

    // Breadth-first traversal
    public void preOrderTraversal(Node node) {
        if (node != null) {
            System.out.println(node.key + " " + node.data);
            this.preOrderTraversal(node.leftChild);
            this.preOrderTraversal(node.rightChild);
        }
    }

    public void inOrderTraversal(Node node) {
        if (node != null) {
            this.inOrderTraversal(node.leftChild);
            System.out.println(node.key + " " + node.data);
            this.inOrderTraversal(node.rightChild);
        }
    }

    public void postOrderTraversal(Node node) {
        if (node != null) {
            this.postOrderTraversal(node.leftChild);
            this.postOrderTraversal(node.rightChild);
            System.out.println(node.key + " " + node.data);
        }
    }

    public static void main(String[] args) {
        BinaryTree binaryTree = new BinaryTree();

        binaryTree.addNode(50, "Jim");
        binaryTree.addNode(25, "James");
        binaryTree.addNode(15, "Joe");
        binaryTree.addNode(30, "Jeff");
        binaryTree.addNode(75, "Jerry");
        binaryTree.addNode(80, "Joel");
        binaryTree.addNode(85, "Jack");

        binaryTree.preOrderTraversal(binaryTree.root);
        System.out.println("*************\n");
        binaryTree.inOrderTraversal(binaryTree.root);
        System.out.println("*************\n");
        binaryTree.postOrderTraversal(binaryTree.root);

        System.out.println("*************\n");
        binaryTree.findNode(85);

        System.out.println("*************\n");
        binaryTree.remove(85);
        binaryTree.preOrderTraversal(binaryTree.root);

        System.out.println("*************\n");
        binaryTree.addNode(85, "Jack");
        binaryTree.reverse(binaryTree.root);
        binaryTree.preOrderTraversal(binaryTree.root);

        System.out.println("*************\n");
        binaryTree.reverse(binaryTree.root);
        Node closestNode = binaryTree.findNodeWithClosestKey(32);
        System.out.println(closestNode.key + " " + closestNode.data);
        binaryTree.addNode(31, "Johnny");
        closestNode = binaryTree.findNodeWithClosestKey(32);
        System.out.println(closestNode.key + " " + closestNode.data);
    }
}
